[Bug]: coverage report failed "make cov"

### Contact Details

_No response_

### What happened?

should generate coverage report

see readme
## Coverage Reporting

```bash
# Generate coverage report
make cov

# View HTML report (if genhtml is installed)
open coverage/index.html
```


### Version

main/latest

### What browsers are you seeing the problem on?

_No response_

### Relevant log output

```shell
make cov
RUSTFLAGS="-Zinstrument-coverage" LLVM_PROFILE_FILE="cov-%p-%m.profraw" \
  cargo test
error: failed to run `rustc` to learn about target-specific information

Caused by:
  process didn't exit successfully: `/home/ci/.rustup/toolchains/1.86.0-aarch64-unknown-linux-gnu/bin/rustc - --crate-name ___ --print=file-names -Zinstrument-coverage --crate-type bin --crate-type rlib --crate-type dylib --crate-type cdylib --crate-type staticlib --crate-type proc-macro --print=sysroot --print=split-debuginfo --print=crate-name --print=cfg -Wwarnings` (exit status: 1)
  --- stderr
  error: the option `Z` is only accepted on the nightly compiler

  help: consider switching to a nightly toolchain: `rustup default nightly`

  note: selecting a toolchain with `+toolchain` arguments require a rustup proxy; see <https://rust-lang.github.io/rustup/concepts/index.html>

  note: for more information about Rust's stability policy, see <https://doc.rust-lang.org/book/appendix-07-nightly-rust.html#unstable-features>

  error: 1 nightly option were parsed

make: *** [Makefile:25: cov] Error 101
```

### Code of Conduct

- [x] I agree to follow this project's Code of Conduct

--- a/.devcontainer/Dockerfile
+++ b/.devcontainer/Dockerfile
@@ -1,11 +1,11 @@
 # .devcontainer/Dockerfile
 FROM debian:bookworm-slim
 
-# Basics + build toolchain + aarch64 cross + qemu emulation
+# Basics + build toolchain + aarch64 cross + qemu emulation + lcov for HTML coverage reports
 RUN set -e; \
       apt-get update; \
       apt-get install -y --no-install-recommends \
-        ca-certificates curl build-essential pkg-config qemu-user qemu-user-static binfmt-support; \
+        ca-certificates curl build-essential pkg-config qemu-user qemu-user-static binfmt-support lcov; \
       arch="$(dpkg --print-architecture)"; \
       if [ "$arch" = "amd64" ]; then \
         apt-get install -y --no-install-recommends \

--- a/Makefile
+++ b/Makefile
@@ -1,4 +1,4 @@
-.PHONY: check test test-emu cov clean install-deps
+.PHONY: check test test-emu cov cov-grcov clean install-deps
 
 # Default target
 all: check test
@@ -20,17 +20,42 @@ test-nextest:
 test-emu:
 	TARGET=aarch64-unknown-linux-gnu ./scripts/test-emu.sh
 
-# Run coverage analysis
+# Run coverage analysis using cargo-llvm-cov
 cov:
-	RUSTFLAGS="-Zinstrument-coverage" LLVM_PROFILE_FILE="cov-%p-%m.profraw" \
-	  cargo test
-	grcov . -s . -t lcov --binary-path ./target/debug/ -o lcov.info --ignore "/*" --ignore "target/*"
-	@echo "Coverage report generated in lcov.info"
+	@echo "[cov] Ensuring nightly toolchain is installed..."
+	@rustup toolchain install nightly --component llvm-tools-preview 2>/dev/null || true
+	@echo "[cov] Ensuring cargo-llvm-cov is installed..."
+	@cargo install cargo-llvm-cov --quiet || true
+	@echo "[cov] Cleaning previous coverage artifacts..."
+	@rm -f cov-*.profraw lcov.info
+	@rm -rf coverage/
+	@echo "[cov] Running tests with coverage instrumentation..."
+	@cargo +nightly llvm-cov --workspace --all-features --lcov --output-path lcov.info --ignore-filename-regex '(/.cargo/|/tests/.*/snapshots/)'
+	@echo "[cov] Coverage report generated in lcov.info"
 	@if command -v genhtml > /dev/null 2>&1; then \
-		genhtml -o coverage lcov.info && echo "HTML coverage report generated in coverage/"; \
+		echo "[cov] Generating HTML coverage report..."; \
+		genhtml lcov.info --output-directory coverage --branch-coverage && echo "[cov] HTML coverage report generated in coverage/"; \
 	else \
-		echo "Install lcov (genhtml) to generate HTML coverage report"; \
+		echo "[cov] Install lcov (genhtml) to generate HTML coverage report"; \
 	fi
+
+# Alternative coverage using grcov (legacy approach - commented out by default)
+# Uncomment and use 'make cov-grcov' if you prefer the old grcov-based workflow
+# cov-grcov:
+# 	@echo "[cov-grcov] Using legacy grcov approach (requires nightly)..."
+# 	@rustup toolchain install nightly --component llvm-tools-preview 2>/dev/null || true
+# 	@cargo install grcov --quiet || true
+# 	@rm -f cov-*.profraw lcov.info
+# 	@rm -rf coverage/
+# 	@RUSTFLAGS="-Zinstrument-coverage" LLVM_PROFILE_FILE="cov-%p-%m.profraw" \
+# 	  cargo +nightly test
+# 	@grcov . -s . -t lcov --binary-path ./target/debug/ -o lcov.info --ignore "/*" --ignore "target/*"
+# 	@echo "[cov-grcov] Coverage report generated in lcov.info"
+# 	@if command -v genhtml > /dev/null 2>&1; then \
+# 		genhtml -o coverage lcov.info && echo "[cov-grcov] HTML coverage report generated in coverage/"; \
+# 	else \
+# 		echo "[cov-grcov] Install lcov (genhtml) to generate HTML coverage report"; \
+# 	fi
 
 # Clean build artifacts
 clean:

